libs/entitlement/asm-provider/src/main/java/org/elasticsearch/entitlement/instrumentation/impl/InstrumenterImpl.java

```diff
@@ -82,13 +91,63 @@ static ClassFileInfo getClassFileInfo(Class<?> clazz) throws IOException {
         return new ClassFileInfo(fileName, originalBytecodes);
     }
 
+    private enum VerificationPhase {
+        BEFORE_INSTRUMENTATION,
+        AFTER_INSTRUMENTATION
+    }
+
+    private static String verify(byte[] classfileBuffer) {
```

Comment:
Can we do this as an assert so we don't need a different system property? Then it will automatically be on during tests.

Reply:
I would avoid it: I think it is OK to have this turned on for our tests, but there are JDK classes that fail verification regularly, and verification is not lightweight/may have side effects (change class loading order). 
ASM is not as bad as ClassFile (I had to revert that, as it was breaking some classes by loading them too agressively/in the wrong loader).

Reply:
I think about this as an instrument for our diagnostics; it could even have been some commented out code  that we could un-comment if we need to do again the kind of investigation we had to do (but I really dislike commented out code, and the ability to turn it on on already compiled binaries is appealing)

Reply:
If some JDK classes fail verification, does that mean we can't enable this in production? If that's true, we probably shouldn't bother with the system property in the first place.

Reply:
> If some JDK classes fail verification

Some JDK classes fail verification as they are (_before_ instrumentation), that's why I don't think we can have this as an assertion, but only in specific tests and only logging.

Reply:
> Some JDK classes fail verification as they are

That seems like something is very wrong then. JDK classes shouldn't be failing verification? @ChrisHegarty ?

Reply:
> does that mean we can't enable this in production?

If we keep the current implementation, which just logs, we can. And the sysprop is to enable it on demand (e.g. on our IT tests, or in cases like the other day where we need to do some deeper investigation)

Reply:
I investigated this a bit more: the JDK classes that fail verification fail _only_ when we try to verify them during transformation. The failure is a pretty specific `ClassCircularityError`. We probably hit a corner case where verification try to load classes (for verification purposes) while we are transforming them.
This means we cannot enable this "by default", via assert or similar. 
We agreed to keep the sysprop to enable it "on demand" in case we need to investigate, but also enable it if we hit a VerifyError: we enable verification and try to transform again to collect details on what fails and why.
I'll push a change shortly.